### DISCUSSION TITLE
New RFC of defineEmits in vue 3.3 ?

### DISCUSSION THREAD

**Discussion by edwinhuish** [vuejs/rfcs#500] [RFC Discussions]
I saw this in twitter [More ergonomic defineEmits() with types coming in 3.3!](https://twitter.com/youyuxi/status/1641403989026820098?s=20),

![image](https://user-images.githubusercontent.com/5207925/233234426-a4b5021f-3c3c-453a-995b-cffb1293f27a.png)

**But... why we just do it like this:** 
```ts
const emit = defineEmits<{
     foo: (id: string) => any;
     bar: (...args: any[]) => any;
}>();
```

**Comment by yyx990803** [comment]
Because the return type is irrelevant in most cases. Why would you prefer a more verbose syntax?

**Reply by edwinhuish** [reply]
Emits are functions, the definition should be function like. If someone like to save words in typing, he should make/find a better IDE for help. It's the IDE's work. Not human.

Also, I don't know how will you type in this case?

```ts
const emit = defineEmits<{
  (e: 'foo', name:string, ...args: any[]): any;
}>(); 
```

**Reply by yyx990803** [reply]
If you want actual function types, you should stick to the original syntax (which is still available). The original syntax is the actual type of the returned `emit` function.

The new syntax is introduced purely for convenience - i.e. shorter syntax is the priority. Your proposed syntax is **neither the actual function type nor the most succinct**.

> Also, I don't know how will you type in this case?

Tuple types can also have rest spread:
```ts
const emit = defineEmits<{
  foo: [name:string, ...args: any[]]
}>()
```

**Reply by edwinhuish** [reply]
I think programing should follow the KISS principles. Any other programmer who just know how to use JavaScript, can know what are you coding easily. Not the various strange ways of writing.

Yes, I can choose the original syntax. If this RFC is the third part library, I will say nothing, it just the personal choice.

> i.e. shorter syntax is the priority

In case of shorter syntax, the more important is follow the KISS principles. Your code is something like below, but even worse, because it breaks away from the original semantics.

> This code is from [Airbnb JavaScript Style Guide](https://github.com/airbnb/javascript#control-statements--value-selection)
```ts
// bad
!isRunning && startRunning();

// good
if (!isRunning) {
  startRunning();
}
```

> Tuple types can also have rest spread

So.... In your syntax, The 2 code below are same?
```ts
const emit = defineEmits<{
  foo: [...args: any[]];
}>();


const emit = defineEmits<{
  foo: any[];
}>();
```

**Reply by darrylnoakes** [reply]
My 2 cents re verbosity:
It's not so much about reducing typing or the number of characters, but allowing the developer to focus on what actually matters.
The important information is the event name and the data it can have. The fact that it's a function isn't the point, nor is its return type, so why specify things that are irrelevant?

**Reply by edwinhuish** [reply]
> My 2 cents re verbosity: It's not so much about reducing typing or the number of characters, but allowing the developer to focus on what actually matters. The important information is the event name and the data it can have. The fact that it's a function isn't the point, nor is its return type, so why specify things that are irrelevant?

All you're talking about are the IDE's job. The code is not only for you to understand at that moment, but also for other collaborating programmers, even for your in the future.

That is the reason why we can't accept meaningless variable names, although it is easy to write much shorter.

In my opinion, this RFC is worse than the meaningless variable names....

**Reply by darrylnoakes** [reply]
Just an example of what I mean:

```typescript
// PLAIN

type Line = {
  type: "line"
  value: string
}

type Section = {
  type: "section"
  value: Line[]
}

// WITH HELPER

type Node<T extends string, V> = {
  type: T
  value: V
}

type Line = Node<"line", string>
type Section = Node<"section", Line[]>
```

All I've done is _abstract away the **common syntax**_.
Which would you choose?

**Reply by edwinhuish** [reply]
Your demo is wrong. 

Your code is normal ts generics type, it makes sence.

If vue strong enough to invent a new script type like "VueScript", it's very good news, you can invet everythings you want for "VueScript", because it's NOT javascript or typescript. Otherwise, I think it's better to follow the javascript/typescript rules.

Same question for you, which do you prefer?  Shorter typing or variable name with meaning & IDE helper?
![image](https://user-images.githubusercontent.com/5207925/235304983-a5d3707c-5b1c-4ae3-811d-948780b7621b.png)

**Reply by darrylnoakes** [reply]
It is following "TS rules": it's a normal generic type.

But, instead of specifying full function types, you only need to specify the name and arguments for each (`[id: string]` is the parameters for a function of the form `(id: string) => any`).

**Reply by jacekkarczmarczyk** [reply]
> Same question for you, which do you prefer? Shorter typing or variable name with meaning & IDE helper?

In the RFC you have names with meaning and IDE support, AND it's short.
If you're objetion is that the generic type you provide to `defineEmits` is not really the type of the actual `emit` function then neither is your proposal, you don't use it like `emit.foo(arg1, arg2)`

**Reply by darrylnoakes** [reply]
> > My 2 cents re verbosity: It's not so much about reducing typing or the number of characters, but allowing the developer to focus on what actually matters. The important information is the event name and the data it can have. The fact that it's a function isn't the point, nor is its return type, so why specify things that are irrelevant?

> All you're talking about are the IDE's job. The code is not only for you to understand at that moment, but also for other collaborating programmers, even for your in the future.

So I should specify irrelevant info, just because my IDE can autocomplete it for me? Contrived extreme case: do you want me to write an extension that inserts a comment with a timestamp every time you create a new line?

**Comment by Demivan** [comment]
Example by @edwinhuish just makes sense. I can't explain why.
This was the syntax that I first tried with generic `defineEmits`, before failing and checking the docs.

Funny thing, this was just suggested to me by Copilot:

```ts
	const emit = defineEmits<{
		'update:modelValue': (value: boolean) => void
	}>()
```

**Reply by edwinhuish** [reply]
Me too....

I don't know how it actually works, but I think the code should be something like:
```ts
const emit = (eventName, ...args) => {
  const callback = $getEmitCallback(eventName);
  if(callback){
    callback(...args);
  }
}
```

**Reply by yyx990803** [reply]
Do you realize that with the type

```ts
const emit = defineEmits<{
  'update:modelValue': (value: boolean) => void
}>()
```

`emit` needs to work like this:

```ts
emit['update:modelValue'](true)
```

Which is obviously wrong. It "breaks away from the original semantics" (in your own words).

If you want the type to match exactly how `emit` works, it should be:

```ts
const emit = defineEmits<{
  (event: 'update:modelValue', value: boolean) => void
}>()
```

This is the original syntax, which is still available.

---

If we are introducing a new syntax that doesn't represent the accurate type of `emit`, then it needs to be as short as possible to be justified. The key: tuple syntax does exactly that.

Your proposed syntax (1) looks like a function, but represents the wrong type; and (2) is more verbose as it requires the useless return type.

I think I've made my points pretty clear here.

**Reply by edwinhuish** [reply]
In this way: 
```ts
const emit = defineEmits<{
  click: (value: boolean) => void
}>()
```

We trigger it in the component:
```ts
emit('click', true)
```

We can easily imagine the `emit` is equal to 
```ts
const emit = (eventName, ...args) => {
  const callback = $getEmitCallback(eventName);
  if(callback){
    callback(...args);
  }
}
```

And we know clearly when we use this component should pass a `Function`, not `Array`
```html
<DemoComponent @click="(val) => clicked = val" />
```

Not `Array` which will confused by the new RFC:
```html
<DemoComponent @click="[val]" />
```

Also, if you want to trigger emit exactly match the definition, I think it's OK to support a new RFC:
```ts
emit.click(true)
```
WHY NOT??

**Reply by edwinhuish** [reply]
New RFC:
```ts
const emit = defineEmits<{
  click: [value: boolean]
}>()
```

My point:
```ts
const emit = defineEmits<{
  click: (value: boolean) => void
}>()
```

The letter ` => void ` I don't think it's hard to autocomplete by IDE...

**Reply by edwinhuish** [reply]
When in the component, definition as RFC:
```ts
const emit = defineEmits<{
  click: [value: boolean]
}>()
```

When we try to undenstand how it work:
```ts
emit('click', true)
```

The train of thought should be like follow:
```ts
const emit = (eventName, ...args) => {

  // HERE WILL RETURN A ARRAY AS DEFINITION.
  // WE DON'T UNDERSTAND WHY AND HOW IT GOING TO WORK.
  // THIS IS WHY I SAY "breaks away from the original semantics"
  const arr = $getEmitVal(eventName);
  // ....
}
```

**Reply by darrylnoakes** [reply]
> The train of thought should be like follow:
> ```typescript
> const emit = (eventName, ...args) => {
> 
>   // HERE WILL RETURN A ARRAY AS DEFINITION.
>   // WE DON'T UNDERSTAND WHY AND HOW IT GOING TO WORK.
>   // THIS IS WHY I SAY "breaks away from the original semantics"
>   const arr = $getEmitVal(eventName);
>   // ....
> }
> ```

Why? It would be the same as you said originally:

```typescript
const emit = (eventName, ...args) => {
  const callback = $getEmitCallback(eventName);
  if (callback) {
    callback(...args);
  }
}
```

The array type in the RFC is simply the parameter list of the event callback.

**Reply by darrylnoakes** [reply]
Why would anyone think they should pass an array?
Your examples make me think you are treating the type as the type of the callback.
**The type is not the type of what must be bound.**
Currently, the full generic type passed is the exact type of the `emit` function itself. With this RFC, it will instead be a map of event names to the data they have, with the data being in the form of a tuple (as used for parameter lists).

**Reply by edwinhuish** [reply]
I think you should ask someone who know typescript but never use vue before, and let him write the code whithout explain to him.

You will know what is "Keep It Simple and Stupid" mean.

**Reply by darrylnoakes** [reply]
Then they can use the original syntax; it's quite understandable, just rather verbose. What's the issue?

The original syntax is simply the type of the `emits` function. The new one is a map of event name to event data.
What's the advantage of your proposed syntax?

I see this:

```typescript
type Original = {
  (e: NAME, DATA_ITEM) => any
  (e: NAME, ...args: DATA_LIST) => any
}

type RFC = {
  NAME: [DATA_ITEM]
  NAME: DATA_LIST
}

type Yours = {
     NAME: (DATA_ITEM) => any
     NAME: (...args: DATA_LIST) => any
}
```

I don't see why the proposed syntax doesn't make sense.
